### DISCUSSION TITLE
Closing the loop on stepper motors.

### DISCUSSION THREAD

**Discussion by Squeack** [grblHAL/core#110] [Q&A]
Moved from https://github.com/terjeio/grblHAL/discussions/356

Stepper motor control is open-loop and cheap. Servos are closed-loop and expensive. I am planning on partly closing the loop on my stepper driven system by feeding the step and directions outputs from the GrblHAL controller (Teensy based breakout board) to both the stepper drivers and to a separate position tracker based on a Raspberry Pi Pico. The three movement leadscrews will be linked via timing belts to 200PPR rotary encoders which will feed back to the Pico. After allowing a ratio for micro-stepping the Pico will then raise an error if the driver instructions and position feedback do not match within a given tolerance.
This has to be communicated to the GrblHAL controller somehow. What would be best? Triggering the door open switch? Does GrblHAL react to the limit switch inputs for the A or B axis even if compiled to handle only three axes?

Terjeio replied in old discussion forum:
Suggestion to use motor fault pin in newer versions. A & B limit switch pins do not respond for 3 axis mills, but can be repurposed.

**Comment by Squeack** [comment]
Motor fault does not seem to be fully supported across all platforms yet. Within the Teensy driver there is an '#if MOTOR_FAULT_ENABLE <then> #define MOTOR_FAULT_PIN n' within GRBLHAL2000_map.h, but MOTOR_FAULT_ENABLE does not appear in any other place, so is never defined.

Following the trail for the safety door activation:
grbl/config.h may define ENABLE_SAFETY_DOOR_INPUT_PIN, although it is commented out by default.
If that is defined, then grbl/driver_opts.h defines SAFETY_DOOR_ENABLE as 1, otherwise as 0.
If SAFETY_DOOR_ENABLE is defined as non-zero then [boardname]_map.h defines SAFETY_DOOR_PIN as the pin number.

Along the same lines:
grbl/config.h should allow defining ENABLE_MOTOR_FAULT_INPUT_PIN.
grbl/driver_opts.h should set MOTOR_FAULT_ENABLE to 1 or 0 accordingly.
If MOTOR_FAULT_ENABLE is non-zero then [boardname]_map.h should define MOTOR_FAULT_PIN.

Within the iMXRT1062 (Teensy) driver, driver.c has multiple functions which make use of SAFETY_DOOR_PIN, but nothing that makes use of MOTOR_FAULT_PIN. grbl/state_machine.c has a STATE_SAFETY_DOOR, but no STATE_MOTOR_FAULT or similar. grbl/system.c uses the safety door status or signal in the control_interrupt_handler(), disable_lock() and go_home(), but only uses signals.motor_fault in control_interrupt_handler() and go_home(), but not disable_lock().

Since STATE_SAFETY_DOOR is likely to be a temporary state that can be resumed from, but STATE_MOTOR_FAULT is likely to be terminal, they probably need to be handled differently, but the core support for motor faults does not seem to be complete yet.

Is this an active area of development?

**Comment by terjeio** [comment]
> Motor fault does not seem to be fully supported across all platforms yet. Within the Teensy driver there is an '#if MOTOR_FAULT_ENABLE #define MOTOR_FAULT_PIN n' within GRBLHAL2000_map.h, but MOTOR_FAULT_ENABLE does not appear in any other place, so is never defined.

Correct, it has to be added to the input definitions array as a minimum.

> grbl/config.h may define ENABLE_SAFETY_DOOR_INPUT_PIN, although it is commented out by default.

I have kept this for backward compatibility - I want to rename, or perhaps remove it, it as pin definitions do _not_ belong in the core. Signal handling functionality does,

> Along the same lines:
grbl/config.h should allow defining ENABLE_MOTOR_FAULT_INPUT_PIN.
grbl/driver_opts.h should set MOTOR_FAULT_ENABLE to 1 or 0 accordingly.
If MOTOR_FAULT_ENABLE is non-zero then [boardname]_map.h should define MOTOR_FAULT_PIN

No, see above. It should be enough to define the pin in the map file. Or perhaps also by adding a plugin that claims an aux input.

> ... but the core support for motor faults does not seem to be complete yet.

Again correct, currently the signal resets the controller so it is pretty harsh. It can surely be improved upon.

> Is this an active area of development?

Not really, but can be made active again if a reasonable functionality can be agreed upon.

This discussion could be what is needed to get it completed?

**Comment by andrewmarles** [comment]
The motor fault signal does require an additional (ideally interrupt capable) input pin on the MCU, and those are always in very short supply.  Of course now that I say that I realize that adding a polling check to the (core) motor fault signal in stepper.c would actually be pretty effective.  At least on platforms like STM32 it would be helpful to poll it since it is pretty easy to fill up all of your interrupt pins on that chip.

To me it seemed that the action taken in the event of motor fault was appropriate (effectively e-stop) since by definition position has been lost if the motor faults and homing is required.  It is useful to the user to have the motor fault code appear rather than e-stop so that it is immediately obvious why the system has halted, so it would be good for us to start exposing this signal in more of the drivers.  I guess if @Squeack with your implementation, if you are able to somehow correlate received step pulses with the encoder position you might be able to keep track of position in the event of following error and use the motor fault to signal the core to pause rather than halt so that the motor could catch up?

**Reply by Squeack** [reply]
@andrewmarles - did you mean have the encoder send a correction back to the controller? I had not thought about that. I do not think that a pause would allow the motor to just catch up, since the fault could be dropped steps or a driver failure. For dropped steps, it might be possible for boards that support a second UART, or primary connection via network and encoder to UART, to trigger HOLD, then feed back to the correct position using move relative, then release HOLD, but this is far more than I was considering. However, my main reason for building this functionality is that I sometimes get a (cheap Chinese) driver stop completely. The fault LED stays dark, but the power LED goes out even though it is still getting power. This is an unrecoverable error as position is lost, so should be treated as equivalent to ESTOP. If not stopped then the router remains stationary on that axis but contines on the others, so usually cuts straight through something it should not. Turning everything off then on gets it all up and running, until it goes wrong again an unpredictable time later. I may just put a relay across (if NO) or inline with (if NC) the ESTOP button and trigger that from the encoder tracker. I know my limit switches are NC, but would have to check the front panel buttons.

It would make sense to have pure signal handling within the core and signal configuration in the hardware driver level. Looking at the use of ENABLE_SAFETY_DOOR_INPUT_PIN: if this is defined within grbl/config.h then SAFETY_DOOR_PIN ends up being defined by [boardname]_map.h and likewise if not defined, so all mentions of ENABLE_SAFETY_DOOR_INPUT_PIN (which is just #ifdef blocks) within the core grbl code can be replaced with SAFETY_DOOR_PIN instead. SAFETY_DOOR_ENABLE is mostly just a bridge defined within grbl/driver_opts.h to enable the _PIN definition. This can be simplifed to just define SAFETY_DOOR_PIN within [boardname]_map.h or leave it undefined and drop ENABLE_SAFETY_DOOR_INPUT_PIN entirely. SAFETY_DOOR_ENABLE is also used within grbl/pin_bits_masks.h in #if statements, so is looking for a value, and is sometimes a compound check with something else as well. These could change to #ifdef SAFETY_DOOR_PIN and any compound checked changed to a nested check.
This would leave only SAFETY_DOOR_PIN to be defined as the pin number or left undefined if not used, at only the driver level, and the handling within the core. MOTOR_FAULT_PIN or OTHER_FEATURE_PIN could be handled the same way. ENABLE_FEATURE_INPUT_PIN and FEATURE_ENABLE would no longer be used at all, so as FEATURE_PIN is both an enablement toggle if defined/undefined  and a location from its defined value, it might be renamed to something like FEATURE_ENABLED_ON_PIN.

Bear in mind that I have only looked at the iMXRT1062 driver, so there may be other uses of these #defines within other drivers that I do not know about.

As an aside, I have also recently replaced the unreliable 'enhanced TB6600', aka S109AFTG, drivers with StepperOnline DM556T drivers. The motors are now much quieter and can support a higher maximum speed, but they are currently unusable for accurate work. Whenever they start from stationary you can hear the motor start up, but it moves a random short distance in a random direction, pauses very briefly, then continues as it should. It is as if the driver, which is on 16x microsteps, is resetting to the nearest whole step at the start of each movement. This leaves the end of the movement in the wrong place, or gets multi-axis movements out of sync. I need to put a 'scope on the step and direction pins to see if the fault lies in the driver or controller, but getting inside the controller box while it is in an operatable state will require some rearrangement. Any suggestions to possible causes or fixes would be welcome.

**Reply by Squeack** [reply]
Just a quick additional thought: The above simplification of the FEATURE_PIN usage might not work with the current code header inclusion order. If the core is built without a link to the hardware layer then FEATURE_PIN would not have been defined yet. Various #includes might need rearranging.

**Reply by terjeio** [reply]
> This can be simplifed to just define SAFETY_DOOR_PIN within [boardname]_map.h or leave it undefined and drop ENABLE_SAFETY_DOOR_INPUT_PIN entirely.

I guess I'll drop ENABLE_SAFETY_DOOR_INPUT_PIN from the core and move it as an option to _my_machine.h_. For that I have to leave safety signal handling always compiled in in the core, that may require some thought.

> This leaves the end of the movement in the wrong place, or gets multi-axis movements out of sync. I need to put a 'scope on the step and direction pins to see if the fault lies in the driver or controller, but getting inside the controller box while it is in an operatable state will require some rearrangement. Any suggestions to possible causes or fixes would be welcome.

Disable stepper disable when idle helps? Set $1 to 255, or $37 to a mask for the motors you want to keep enabled.

> Just a quick additional thought: The above simplification of the FEATURE_PIN usage might not work with the current code header inclusion order. If the core is built without a link to the hardware layer then FEATURE_PIN would not have been defined yet. Various #includes might need rearranging.

The core does not know anything about the #define symbols created at the driver level - and should not. The driver advertises capabilities to the core via the HAL structure, either by setting optional [function pointers](https://github.com/grblHAL/core/blob/9a904127b441ca3d08c38155190f1f235b8ef4dc/hal.h#L556-L669) or by setting [capability flags](https://github.com/grblHAL/core/blob/9a904127b441ca3d08c38155190f1f235b8ef4dc/hal.h#L43-L73). There _are_ files in the core that may suggest otherwise but these are only referenced by driver or plugin code - added to the core for convenience.

**Reply by Squeack** [reply]
My comment about having to change the #includes was because grbl/config.h current may define ENABLE_SAFETY_DOOR_INPUT_PIN, and has optional code based on that define. If the three SAFETY_DOOR defines are condensed into one, which is at driver level, then how does the optional code in the core get enabled or disabled? Capability flags (which I had not dug deep enough into the code to know about) look like a good possibility. It is run-time rather than compile-time, which might have made a difference on the tightly constrained original GRBL running on a ATmega, but is likely to be acceptable on a more capable platform.

**Reply by Squeack** [reply]
Thanks for the suggestion of setting $1 to 255. It worked.

**Reply by terjeio** [reply]
FYI `#define ENABLE_SAFETY_DOOR_INPUT_PIN` has now been [removed from the core](https://github.com/grblHAL/core/blob/master/changelog.md).
